### PR TITLE
Naheed MediaRanker Edges

### PR DESCRIPTION
**Pull request by arangn**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a method to sort works based on votes in my works model
Describe how you approached testing that model method. What edge cases did you come up with? | I tested the method to make sure it sorted works based on votes, and also tested whether the works show up if there are none as an edge case
What are `session` and `flash`? What is the difference between them? | Session is to record the current user, while flash is used to display a message to the user.
Describe a controller filter you wrote. | I created a find work filter that finds a record based on the id
What was one thing that you gained more clarity on through this assignment? | How everything talks to each other, how to do a rails project start to finish, how to copy paste code without feeling guilty
What is the Heroku URL of your deployed application? | https://media-ranker-naheed.herokuapp.com
Do you have any recommendations on how we could improve this project for the next cohort? | More time pls

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    @user = User.find_by(username: name)
+
+    if @user
+      flash[:success] = "Successfully logged in as #{name}"
+      session[:user_id] = @user.id
+      redirect_to root_path
+    else
+      @user = User.new(username: name)
+      @user.save
+      redirect_to root_path
```
Comment (droberts-sea) [inline app/controllers/sessions_controller.rb:17]: When you create a new user you're not saving their ID in the session. That means that a new user has to enter their name twice to log in: once to create the record, and once to set their user ID in the session.

File: app/controllers/works_controller.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  end
+
+  def create
+    filtered_params = work_params()
+    @work = Work.new(filtered_params)
+    save_success = @work.save
+    if save_success
+      redirect_to works_path
+    else
+      render :new
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb:43]: It would be good to put a message in the flash in both these cases, either "successfully created work" or "could not create work".

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :votes
+  validates :username, presence: true, uniqueness: true
+
+  def self.voted_works(user_id)
+    works = User.where(id: user_id)
+    return works
+  end
```
Comment (droberts-sea) [inline app/models/user.rb:8]: This method returns a list of (one) user, not a list of works. Why not use a `has_many :through` relationship here?

File: app/models/vote.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
+  validates :work, uniqueness: { scope: :user }
+  validates :user, uniqueness: { scope: :work }
```
Comment (droberts-sea) [inline app/models/vote.rb:5]: Good work getting this tricky `uniqueness` scope figured out! However, I think you only need one or the other of these lines, not both.

File: app/models/work.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  def self.sort_works(category)
+    works = Work.where(category: category)
+    sorted_works = works.sort_by {|work| work.votes.count}
+    return sorted_works.reverse
+  end
+
+  def self.list_top_ten(category)
+    works = Work.sort_works(category)
+    return works[0..9]
+  end
```
Comment (droberts-sea) [inline app/models/work.rb:14]: I love the way these model methods build on each other, each adding a small bit of functionality. This is a great example of functional decomposition.

File: app/views/home/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    <p class="spotlight__description"><%= Work.find_top_work.votes.count %> votes - <%= Work.find_top_work.description %>
+    </p>
+  </section>
+  <hr class="root__hr" />
+  <section class="top-ten__container">
+  <section class="top-ten__list-container">
+    <h2 class="top-ten__header">
+      Top Movies
+    </h2>
+    <ul class="list-group top-ten__list">
```
Comment (droberts-sea) [inline app/views/home/index.html.erb:17]: You have the same code to show a list of works repeated 3 times. Could you use a view partial or a loop to DRY this up?

File: config/routes.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  root 'home#index'
+
+  resources :works
+  post '/works/:id/vote', to: 'works#vote', as: 'vote_work'
+
+  resources :sessions, only: [:new, :create]
+  post '/sessions/logout', to: 'sessions#logout', as: 'logout'
+
+  resources :users
+
```
Comment (droberts-sea) [inline config/routes.rb:12]: I don't think you need all 7 RESTful routes for users. This site should only use `index` and `show`.

File: test/models/vote_test.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+require "test_helper"
+
+describe Vote do
+  let(:vote) { Vote.new }
+
+  it "must be valid" do
```
Comment (droberts-sea) [inline test/models/vote_test.rb:6]: You should be testing the uniqueness constraint here! Specific test cases I'd want to see:

- A user can have votes for two different works
- A work can have votes from two different users
- A user cannot vote for the same work twice

File: test/models/work_test.rb [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+    # Assert
+    expect(result).must_equal false
+    expect(@work.errors.messages).must_include :title
+    end
+  end
+  describe "sort_works" do
+    it "sorts works based on votes" do
+      sorted = Work.sort_works("album")
+      first = sorted.first.votes.count
```
Comment (droberts-sea) [inline test/models/work_test.rb:25]: This is a good start to testing for your custom model methods, but there's some interesting edge cases you haven't covered.

For `find_top_work`, I would wonder:

- What happens if there are no works?
- What happens if there are works but no votes?
- What happens if two works have the same number of votes?

Similarly for your `sort_works` and `top_ten` methods, I would ask:

- What if there are no works of that category?
- What if there are less than 10 works?
- What if there's a tie for last place, e.g. works 9, 10 and 11 all have 0 votes?

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Comprehension questions	| yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | some extra routes generated
Views are well-organized (DRY, use of semantic HTML, use of partials) | great semantic HTML, some repetition of view code (see inline)
Errors are reported to the user | some - this could be much more comprehensive
Business logic lives in the models | yes - great work
Models are thoroughly tested, including relations, validations and any custom logic | good start, but many edge cases missing - see inline
**Wave 1 - Media** |
Splash page shows the three media categories | yes
Basic CRUD operations on media are present and functional | I was not able to create a new work, but the other operations worked fine
**Wave 2 - Users and Votes** |
Users can log in and log out | yes (but see inline)
The ID of the current user is stored in the session | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | no
Individual user pages and the user list are present | some - list of votes on the user page isn't quite there
**Optional - Styling** |
Bootstrap is used appropriately | yes
Look and feel is similar to the original | workflow is the same, styling less so
**Overall** |

Good work overall. I was right on the border between yellow and green on this one. Some of the wave 3 functionality is incomplete, but in general this site functions well, and it is clear to me that the learning goals for this assignment were met, especially around working with complex business logic.

I don't have much insight into this, but probably the place where there's the most room for growth is in writing code quickly and efficiently. Part of this is just practice, but if you'd like to discuss tools for working on this I'd be happy to.

There is also some room for improvement around testing, but you're on the right track here. As you work on bEtsy, it might be worthwhile to seek out an instructor or TA to get feedback on test coverage before the deadline.

There are a few places where things could be cleaned up, which I've tried to outline below, but in general I am pretty happy with this submission. Keep up the hard work!
